lib/ansible/modules/cloud/misc/qemu_img.py

```diff
@@ -0,0 +1,222 @@
+        ),
+        supports_check_mode=True,
+        required_if=[
+            ("state", "present", ["size"]),
+        ],
+    )
+
+    result = dict(
+        changed=False,
+    )
+
+    state = module.params['state']
+    dest = module.params['dest']
+    img_format = module.params['format']
+    img_options = ','.join(module.params['options'])
+    size = module.params['size']
+    grow = module.params['grow']
+    shrink = module.params['shrink']
+
+    qemu_img = module.get_bin_path('qemu-img', True)
+
+    if state == 'present':
+        if not os.path.exists(dest):
+            if not module.check_mode:
+                if not img_options:
+                    rc, stdout, stderr = module.run_command('%s create -f %s "%s" %s' % (qemu_img, img_format, dest, size), check_rc=True)
+                else:
+                    rc, stdout, stderr = module.run_command('%s create -f %s -o "%s" "%s" %s' % (qemu_img, img_format, img_options, dest, size), check_rc=True)
+            result['changed'] = True
+        else:
+            try:
+                size_unit = size[-1:]
+                size = size[:-1]
+            except Exception as e:
+                size_unit = 'b'
+            try:
+                units = list(['b', 'k', 'M', 'G', 'T', 'P', 'E', 'Z', 'Y'])
+                unitmultiplier = 1024**units.index(size_unit)
+            except Exception as e:
+                module.fail_json(msg='Unknown size unit : %s' % (size_unit))
+            size_increase = False
+            size_decrease = False
+            if (size[:1] == '+'):
+                size_increase = True
+                size = size[1:]
+            elif (size[:1] == '-'):
+                size_descrease = True
+                size = size[1:]
+            size = int(float(size) * float(unitmultiplier))
+            rc, stdout, stderr = module.run_command('%s info "%s"' % (qemu_img, dest), check_rc=True)
+            current_size = None
+            for line in stdout.splitlines():
+                if 'virtual size' in line:
+                    # virtual size: 5.0M (5242880 bytes)
+                    current_size = int(line.split('(')[1].split()[0])
+            if not current_size:
+                module.fail_json(msg='Unable to read virtual disk size of %s' % (dest))
+            if (grow):
+                newsize = size - current_size
+                if (newsize > 0):
+                    if not module.check_mode:
+                        rc, stdout, stderr = module.run_command('%s resize "%s" %s' % (qemu_img, dest, size), check_rc=True)
+                    result['changed'] = True
+            if (shrink):
+                if (img_format == 'qcow2'):
+                    module.fail_json(msg="qemu-img does not support shrinking qcow2 images")
+                if (size[:1] == '-'):
+                    newsize = current_size - size
+                else:
+                    newsize = current_size - size
```

Comment:
So the above is also wrong, in both cases we subtract ? No need for the if-else than?